File: 733/q4.cpp
#include<bits/stdc++.h>

using namespace std;

//typedef long long int;
#define inf 1e18
#define mod 1000000007
#define pb push_back



int main(int argc, char const *argv[])
{
    int T;
    cin>>T;
    
    for(int t=0; t<T; t++){
        int n;
        cin>>n;

        unordered_map<int,bool> val;
        unordered_map<int,bool> nd;
        unordered_map<int,bool> na;
        int ans[n];

        for (int i = 1; i <=n; i++){
        
            nd.insert({i,true});
            na.insert({i,true});
        }
        
        int a[n];
        for(int i=0; i<n; i++){
            int ai;
            cin>>ai;
            a[i] = ai;
            if(val.find(a[i])==val.end()){
                val[a[i]]=i+1;
                nd.erase(nd.find(i+1));
                na.erase(na.find(a[i]));
                ans[i]=a[i];
            }
        }

        cout<<n-na.size()<<endl;

        for(auto& i: nd){
            bool done =false;
            for (auto& j:na){
                if(i.first!=j.first){
                    ans[i.first-1]=j.first;
                    na.erase(na.find(j.first));
                    done =true;
                    break;
                }
            }
            if(!done){
                if(nd.size()!=1){
                    for (auto& k: nd){
                        if(k.first!=i.first){
                            ans[i.first-1]=ans[k.first-1];
                            ans[k.first-1]=i.first;
                            break;
                        }
                    }
                }
                else{
                    ans[i.first-1]=a[i.first-1];
                    ans[val[a[i.first-1]]-1]=i.first;
                }
            }
        }

        for (int i = 0; i < n; i++){
        
            cout<<ans[i]<<" ";
        }
        cout<<""<<endl;
    }
    return 0;
}

File: 733/q4.py
for _ in range(int(input())):
    n = int(input())
    a = list(map(int,input().split()))
 
    val={}
    nd = {i+1:True for i in range(n)}
    na = {i+1:True for i in range(n)}
    ans = [0]*n
    for i in range(n):
        if(val.get(a[i])==None):
            val[a[i]]=i+1
            nd.pop(i+1)
            na.pop(a[i])
            ans[i]=a[i]
 
    # print(ans,nd,na,val)
    print(n-len(na))
    for i in nd:
        done =False
        for j in na:
            if(j!=i):
                ans[i-1]=j
                na.pop(j)
                done =True
                break
        if(not done):
            if(len(nd)!=1):
                for k in nd:
                    if(i!=k):
                        ans[i-1]=ans[k-1]
                        ans[k-1]=i
                        break
            else:
                ans[i-1]=a[i-1]
                ans[val[a[i-1]]-1]=i
 
    for i in ans:
        print(i,end=" ")
    print("")